x/stake/handler_test.go

```diff
@@ -33,6 +33,78 @@ func newTestMsgDelegate(delegatorAddr, validatorAddr sdk.Address, amt int64) Msg
 
 //______________________________________________________________________
 
+func TestValidatorByPowerIndex(t *testing.T) {
+	validatorAddr, validatorAddr3 := addrs[0], addrs[1]
+
+	initBond := int64(1000000)
+	initBondStr := "1000"
+	ctx, _, keeper := createTestInput(t, false, initBond)
+
+	// create validator
+	msgCreateValidator := newTestMsgCreateValidator(validatorAddr, pks[0], initBond)
+	got := handleMsgCreateValidator(ctx, msgCreateValidator, keeper)
+	assert.True(t, got.IsOK(), "expected create-validator to be ok, got %v", got)
+
+	// verify the self-delegation exists
+	bond, found := keeper.GetDelegation(ctx, validatorAddr, validatorAddr)
+	require.True(t, found)
+	gotBond := bond.Shares.Evaluate()
+	require.Equal(t, initBond, gotBond,
+		"initBond: %v\ngotBond: %v\nbond: %v\n",
+		initBond, gotBond, bond)
+
+	// verify that the by power index exists
+	validator, found := keeper.GetValidator(ctx, validatorAddr)
+	require.True(t, found)
+	pool := keeper.GetPool(ctx)
+	power := GetValidatorsByPowerKey(validator, pool)
+	require.True(t, keeper.validatorByPowerIndexExists(ctx, power))
+
+	// create a second validator keep it bonded
+	msgCreateValidator = newTestMsgCreateValidator(validatorAddr3, pks[2], int64(1000000))
+	got = handleMsgCreateValidator(ctx, msgCreateValidator, keeper)
+	assert.True(t, got.IsOK(), "expected create-validator to be ok, got %v", got)
+
+	// slash and revoke the first validator
+	keeper.Slash(ctx, pks[0], 0, sdk.NewRat(1, 2))
+	keeper.Revoke(ctx, pks[0])
+	validator, found = keeper.GetValidator(ctx, validatorAddr)
+	require.True(t, found)
+	require.Equal(t, sdk.Unbonded, validator.PoolShares.Status)             // ensure is unbonded
+	require.Equal(t, int64(500000), validator.PoolShares.Amount.Evaluate()) // ensure is unbonded
+
+	// the old power record should have been deleted as the power changed
+	assert.False(t, keeper.validatorByPowerIndexExists(ctx, power))
+
+	// but the new power record should have been created
+	validator, found = keeper.GetValidator(ctx, validatorAddr)
+	require.True(t, found)
+	pool = keeper.GetPool(ctx)
+	power2 := GetValidatorsByPowerKey(validator, pool)
+	require.True(t, keeper.validatorByPowerIndexExists(ctx, power2))
+
+	// inflate bunch to increase the value of the bonded shares
```

Comment:
I think this comment is incorrect with the inflation changes - loose unbonded tokens are now what's increasing?